core/payments/client_ledger.go

```diff
@@ -0,0 +1,270 @@
+	if err != nil {
+		var timeMovedBackwardErr *reservation.TimeMovedBackwardError
+		if errors.As(err, &timeMovedBackwardErr) {
+			// this is the only class of error that can be returned from Debit where trying again might help
+			return nil, fmt.Errorf("debit reservation: %w", err)
+		}
+
+		// all other modes of failure are fatal
+		panic(fmt.Sprintf("reservation debit failed: %v", err))
+	}
+
+	if !success {
+		return nil, fmt.Errorf(
+			"reservation lacks capacity for blob with %d symbols (%d bytes), "+
+				"and no on-demand fallback is configured",
+			blobLengthSymbols, blobLengthSymbols*encoding.BYTES_PER_SYMBOL)
+	}
+
+	paymentMetadata, err := core.NewPaymentMetadata(cl.accountID, now, nil)
+	enforce.NilError(err, "new payment metadata: %v", err)
+	return paymentMetadata, nil
+}
+
+// Used by ClientLedger instances where only on-demand payments are configured.
+func (cl *ClientLedger) debitOnDemandOnly(
+	ctx context.Context,
+	now time.Time,
+	blobLengthSymbols uint32,
+	quorums []core.QuorumID,
+) (*core.PaymentMetadata, error) {
+	cumulativePayment, err := cl.onDemandLedger.Debit(ctx, blobLengthSymbols, quorums)
+	enforce.NilError(err, "on-demand debit failed. reservations aren't configured, and the ledger won't become "+
+		"aware of new on-chain deposits without a restart: %v", err)
+
+	paymentMetadata, err := core.NewPaymentMetadata(cl.accountID, now, cumulativePayment)
+	enforce.NilError(err, "new payment metadata: %v", err)
+
+	cl.accountantMetricer.RecordCumulativePayment(cl.accountID.Hex(), cumulativePayment)
+
+	return paymentMetadata, nil
+}
+
+// Used by ClientLedger instances where both reservation and on-demand payments are configured.
+//
+// First tries to pay for a dispersal with the reservation, and falls back to on-demand if the reservation
+// lacks capacity.
+func (cl *ClientLedger) debitReservationOrOnDemand(
+	ctx context.Context,
+	now time.Time,
+	blobLengthSymbols uint32,
+	quorums []core.QuorumID,
+) (*core.PaymentMetadata, error) {
+	// As the client, "now" and the dispersal time are the same. The client is responsible for populating the
+	// dispersal time when constructing the payment header, and it does so with its conception of "now"
+	success, err := cl.reservationLedger.Debit(now, now, blobLengthSymbols, quorums)
+	if err != nil {
+		var timeMovedBackwardErr *reservation.TimeMovedBackwardError
+		if errors.As(err, &timeMovedBackwardErr) {
+			// this is the only class of error that can be returned from Debit where trying again might help
+			return nil, fmt.Errorf("debit reservation: %w", err)
+		}
+
+		// all other modes of failure are fatal
+		panic(fmt.Sprintf("reservation debit failed: %v", err))
+	}
+
+	if success {
+		paymentMetadata, err := core.NewPaymentMetadata(cl.accountID, now, nil)
+		enforce.NilError(err, "new payment metadata: %v", err)
+		return paymentMetadata, nil
+	}
+
+	cl.logger.Infof("Reservation lacks capacity for blob with %d symbols (%d bytes). Falling back to on-demand.",
```

Comment:
When reservation is turned on, are you checking that the reservation is actually >0 and panic'ing if not?

Reply:
perfect thanks